### ISSUE TITLE
Failure to optimise shift+and+boolean conversion pattern to simple comparison

### ISSUE THREAD

**Issue by GabrielRavier** [llvm/llvm-project#55739]
```cpp
int
f3 (int c)
{
  return ((2 << c) & 16) != 0;
}

int
f4 (int d)
{
  return ((16 << d) & 32) == 0;
}

int
f8 (int h)
{
  return ((32 >> h) & 16) == 0;
}
```

These functions can respectively be optimized to `return c == 3;`, `return c != 1;` and `return c != 1`. This transformation is done by GCC, but not by LLVM.

See also [Godbolt comparison](https://godbolt.org/z/969P6849b)

**Comment by bcl5980** [comment]
Maybe we can add some code in foldICmpUsingKnownBits to do this. In this function we already have similar case but the shift const is 1.
```
          // ((1 << X) & 8) == 0 -> X != 3
          // ((1 << X) & 8) != 0 -> X == 3
```

https://alive2.llvm.org/ce/z/a6HVAC

**Comment by bcl5980** [comment]
trying to fix this on the patch:
https://reviews.llvm.org/D126591

**Comment by bcl5980** [comment]
We haven't fix all cases in this issue. So we still can't close the ticket I think

**Comment by nunoplopes** [comment]
Ok, thanks! I got confused by the commit message.

**Comment by bcl5980** [comment]
> Ok, thanks! I got confused by the commit message.

Sorry for the commit message confuse you. I will write commit message more accuracy next commit.

**Comment by bcl5980** [comment]
Fixed by bfde8619355
